### ISSUE TITLE
albumentations can get the same results with torchvision?

### ISSUE THREAD

**Issue by guanfuchen** [conradry/pixpro-with-weights#4]
Great work!!! Some questions.
1. Can albumentations get the same results with torchvision? Can be faster?
2. [this BYOL codebase](https://github.com/yaox12/BYOL-PyTorch) can get the right results with DeepMind
3. Did you recover the results as the paper?

**Comment by conradry** [comment]
Thank you!

1. Albumentations is faster than torchvision. It also makes it very easy to apply the same transforms to multiple images. I use that feature to keep track of the locations of pixels relative to the original image (see grid_x and grid_y in data.py).
2. From a cursory glance, it looks like my code matches most of the details from that BYOL repo.
3. I'm training a model right now and it should finish in ~2 days and then maybe another few days to finetune on COCO/PASCAL. If it works, I should have results to share sometime next week.

**Comment by conradry** [comment]
@guanfuchen, I added pre-trained weights. The results are not as good as those reported in the paper, but they are on par with other unsupervised pre-training methods. My best guess is that the encoder momentum that I used was a little too high for the batch size. I'll try another run later.

**Comment by guanfuchen** [comment]
In fact I use the code for training, but it is slow firstly, only 1000FPS 32 V100 GPU. And the negative loss is confuing for me. I think the acc is slow, may LARS excludes weight decay for bias and bn. What is the model linear acc in ImageNet-1k?

![image](https://user-images.githubusercontent.com/22321977/105671012-00f9e700-5f1d-11eb-8b3b-24cab3156761.png)

**Comment by conradry** [comment]
I'm not sure there's much that can be done about slow training speeds. (Do you know what the bottleneck was? Data loading? Forward or backward pass? etc.) Every forward pass of the model is really 4 forward passes (two through the encoder, two through the momentum encoder). Training on ImageNet with 4 V100s took about 4 days for me, which is the same as when I've trained models with MoCo (200 epochs vs 100 epochs though; the symmetric loss is slower per epoch but needs fewer epochs to converge).

I agree that the negative loss is confusing. I requested a loss plot from the authors to help calibrate my results, but they haven't responded.

I updated the code to exclude batchnorm and bias parameters from LARS and weight decay. It might be responsible for the lower accuracy; I haven't had a chance to test it yet.

https://github.com/conradry/pixpro-with-weights/blob/3581ca0013e2c8f7e3bd52a6c9e565f08b73ed3f/train_backbone.py#L367-L413

I haven't tested on ImageNet yet. I plan to give it a try after training a new model with the LARS corrections.

**Comment by guanfuchen** [comment]
Hi, @conradry  Did you also want to add [the contrastive loss](https://github.com/lucidrains/pixel-level-contrastive-learning/blob/main/pixel_level_contrastive_learning/pixel_level_contrastive_learning.py#L422) with pixpro loss?

![image](https://user-images.githubusercontent.com/22321977/105812786-c27e2e00-5fe9-11eb-899f-2200c8d12dd7.png)

**Comment by conradry** [comment]
I'm holding off on adding the contrastive loss until I'm confident in the baseline Pixel Propagation results. Initially, I just want to reach the 55.1% top-1 accuracy on ImageNet, then I'll add the contrastive loss and aim for the 66.3% accuracy.

**Comment by guanfuchen** [comment]
@conradry , I use some code from your repo, except I use pytorch transformation(album is very slow in my environment), and I use same hyper parameter with you or the paper, also use LARS without bias and bn parameter, now get accuracy gain against your pretrained model in pets finetune, yours top1 acc is 87.44, the model which I train gets 89.76 acc. I am testing the linear acc in Imagenet now. Hope it will give you some tips. At last, great work about the implementation, Thxs!!!

**Comment by conradry** [comment]
Thanks @guanfuchen, it helps to know that excluding the bias and bn parameters improves results. I'm training another model right now and will update the detection results on VOC sometime next week as well.

**Comment by guanfuchen** [comment]
@conradry update the result for ImageNet Linear acc 55.876(just use pixpro), the origin paper acc is 55.1, I think it is almost same.
